hcl/hcl_test.go

```diff
@@ -281,6 +342,225 @@ terramate {
+			input: `
+terramate {
+
+}
+`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{},
+				},
+			},
+		},
+		{
+			name: "empty config block returns empty config",
+			input: `
+terramate {
+	config {}
+}
+`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{
+						RootConfig: &hcl.RootConfig{},
+					},
+				},
+			},
+		},
+		{
+			name: "unrecognized config attribute",
+			input: `
+terramate {
+	config {
+		something = "bleh"
+	}
+}
+`,
+			want: want{
+				err: hcl.ErrMalformedTerramateConfig,
+			},
+		},
+		{
+			name: "empty config.git block",
+			input: `
+terramate {
+	config {
+		git {}
+	}
+}
+`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{
+						RootConfig: &hcl.RootConfig{},
+					},
+				},
+			},
+		},
+		{
+			name: "multiple config blocks - fails",
+			input: `
+terramate {
+	config {}
+	config {}
+}
+	`,
+			want: want{
+				err: hcl.ErrMalformedTerramateConfig,
+			},
+		},
+		{
+			name: "basic config.git block",
+			input: `
+terramate {
+	config {
+		git {
+			branch = "trunk"
+		}
+	}
+}
+`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{
+						RootConfig: &hcl.RootConfig{
+							Git: hcl.GitConfig{
+								Branch: "trunk",
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "all fields set for config.git",
+			input: `
+terramate {
+	config {
+		git {
+			branch = "trunk"
+			remote = "upstream"
+			baseRef = "upstream/trunk"
+			defaultBranchBaseRef = "HEAD~2"
+		}
+	}
+}
+`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{
+						RootConfig: &hcl.RootConfig{
+							Git: hcl.GitConfig{
+								Branch:               "trunk",
+								Remote:               "upstream",
+								BaseRef:              "upstream/trunk",
+								DefaultBranchBaseRef: "HEAD~2",
+							},
+						},
+					},
+				},
+			},
+		},
+	} {
+		testParser(t, tc)
+	}
+}
+
+func TestHCLParserStack(t *testing.T) {
+	for _, tc := range []testcase{
+		{
+			name: "empty stack block",
+			input: `
+terramate {
+	required_version = ""
+}
+stack {}`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{},
+					Stack:     &hcl.Stack{},
+				},
+			},
+		},
+		{
+			name: "multiple stack blocks",
+			input: `stack{}
+stack {}`,
+			want: want{
+				err: hcl.ErrMalformedTerramateConfig,
+			},
+		},
+		{
+			name: "empty name",
+			input: `
+terramate {
+	required_version = ""
+}
+stack {
+	name = ""
+}`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{},
+					Stack:     &hcl.Stack{},
+				},
+			},
+		},
+		{
+			name: "name is not a string - fails",
+			input: `
+terramate {
+	required_version = ""
+}
+stack {
+	name = 1
+}`,
+			want: want{
+				err: hcl.ErrMalformedTerramateConfig,
+			},
+		},
+		{
+			name: "name has interpolation - fails",
+			input: `
+terramate {
+	required_version = ""
+}
+stack {
+	name = "${test}"
+}`,
```

Comment:
maybe we could extract the terramate default blocks in a way that avoids it being repeated on all tests, so tests only shows stack blocks (terramate block seems to always be the same, so kinda unrelated to what is being tested).